### DISCUSSION TITLE
Add support for `or` statement in generic‘s constraints

### DISCUSSION THREAD

**Discussion by huhaorui** [dotnet/csharplang#4737] [Language Ideas] [answered]
In this version of C#, we can constraint the generic‘s type by `where` statement, and set multiple conditions by comma. Howerver, when we want to constraint by multiple optional conditions, the language has no support.

**Comment by yaakov-h** [comment, marked as answer]
Can you please clarify what you mean by "multiple optional conditions"?

**Reply by huhaorui** [reply]
For example, in this situation：
```c#
        public static void Test<T>(this T obj) where T : IEnumerable or ICollection
        {
            //do somethind
        }
```

**Reply by yaakov-h** [reply]
This isn't something that the runtime can currently support. See #399 for further discussion in this area.

**Reply by JustNrik** [reply]
> 
> 
> For example, in this situation：
> 
> ```cs
>         public static void Test<T>(this T obj) where T : IEnumerable or ICollection
>         {
>             //do somethind
>         }
> ```

And how is obj supposed to be treated? What's the benefit of this? Why can't you just use IEnumerable since ICollection inherits from it so you can just pattern match?

**Reply by jnm2** [reply]
Maybe `IReadOnlyCollection<T> or ICollection<T>` is a better example, since each defines Count and one doesn't inherit from the other.

**Reply by smoothdeveloper** [reply]
@huhaorui in your original suggestion, are you looking at code to be monomorphised or to work through reflection?

In case of the later, maybe a feature like [F# Statically Resolved Type Parameters](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/generics/statically-resolved-type-parameters) would be suitable:

* it requires no runtime changes (it monorphises the code at IL, it would be matter of the compiler inlining the method implementation)
* it supports the `or` on generic type parameters

It is a complex and large feature though, as are those that deal with generic and the type system, especially if runtime is going to be involved.

If following F# design, it would require the compiler to embed the metadata about the method, so it can be monomorphised / dispatched appropriately, even crossing the assembly barrier.

**Comment by hez2010** [comment]
This is definitely a feature extremely useful for expressing sum types, especially after introducing SIM. I think C# need either allowing overloading by generic constraints or introducing `or` for generic constraints. Although this would require a runtime change to support it, it really worth. 

For example:

```csharp
T Foo<T>(T x) where T : IBinaryIntegerNumber<T> { ... }
T Foo<T>(T x) where T : IBinaryFloatingNumber<T> { ... }
```

or:
```csharp
T Foo<T>(T x) where T : IBinaryIntegerNumber<T> | IBinaryFloatingNumber<T>
{
    // Then use pattern matching here
}
```
